File: src/pyscf/pyscf_module.py

import warnings
warnings.filterwarnings("ignore", category=UserWarning, module="numpy.core.getlimits")
import numpy as np
from pyscf import gto



def read_xyz(xyz):

    f = open(xyz, 'r')

    lines = f.read().splitlines()

    # nb of atoms
    nbAt = int(lines.pop(0))
    lines.pop(0)

    # list of atoms positions
    list_pos_atom = []
    for line in lines:
        tmp = line.split()
        atom = tmp[0]
        pos = (float(tmp[1]), float(tmp[2]), float(tmp[3]))
        list_pos_atom.append([atom,pos])

    f.close()

    return list_pos_atom


def mol_prop(xyz: str, input_basis: str, charge: int, multiplicity: int, unit: str, cartesian: bool):

    list_pos_atom = read_xyz(xyz)

    mol = gto.M(
        atom = list_pos_atom,
        basis = input_basis,
        charge = charge,
        spin = multiplicity - 1
    )

    mol.unit = unit
    mol.cart = cartesian

    mol.build()

    natoms = mol.natm       # nb of atoms
    nalpha = mol.nelec[0]   # nb of alpha-electrons
    nbeta = mol.nelec[1]    # nb of beta-electrons
    Enuc = mol.energy_nuc() # nuclear energy

    return natoms, nalpha, nbeta, Enuc

File: src/pyscf/pyscf_module.f90
module pyscf_module

    use, intrinsic :: iso_c_binding

    implicit none

    interface

        subroutine call_mol_prop(xyz, input_basis, charge, multiplicity, unit_type, cartesian, &
                                 natoms, nalpha, nbeta, Enuc) bind(C, name="call_mol_prop")

            import c_char, c_int, c_double, c_ptr

            character(kind=c_char), intent(in) :: xyz(*)
            character(kind=c_char), intent(in) :: input_basis(*)
            integer(c_int), intent(in), value  :: charge
            integer(c_int), intent(in), value  :: multiplicity
            character(kind=c_char), intent(in) :: unit_type(*)
            integer(c_int), intent(in), value  :: cartesian
            integer(c_int), intent(out)        :: natoms
            integer(c_int), intent(out)        :: nalpha
            integer(c_int), intent(out)        :: nbeta
            real(c_double), intent(out)        :: Enuc
        end subroutine call_mol_prop

    end interface

end module pyscf_module

